From adfc10233f9be7aad6d44cc4b77d56117c1298b2 Mon Sep 17 00:00:00 2001
From: Loic Poulain <loicx.poulain@intel.com>
Date: Wed, 9 Oct 2013 18:36:30 +0200
Subject: [PORT FROM MAIN]wifi/nl80211: enable wowlan at init

BZ: 143243

Wowlan must be enabled to avoid Wifi disconnection on suspend.
Enable wowlan via nl80211.

Change-Id: I4e9bd4953431d49ded0254b7d263372a7fe7cfb1
Orig-Change-Id: Iedd0b17fe192955c83ca2c253043e88510043638
Category: device enablement
Domain: CWS.WIFI-Bluewoods
Origin: upstream
Upstream-Candidate: no,proprietary
Signed-off-by: Loic Poulain <loicx.poulain@intel.com>
Reviewed-on: http://android.intel.com:8080/137344
Reviewed-by: Trivelly, Jean <jean.trivelly@intel.com>
Reviewed-by: De Cesco, Jonathan <jonathan.de.cesco@intel.com>
Reviewed-by: Bachot, Jean-michel <jean-michel.bachot@intel.com>
Tested-by: Lestienne, ChristopheX <christophex.lestienne@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 src/drivers/driver_nl80211.c |   39 +++++++++++++++++++++++++++++++++++++++
 1 files changed, 39 insertions(+), 0 deletions(-)

diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index f9765fa..c77c9ff 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -3611,6 +3611,43 @@ static void nl80211_destroy_bss(struct i802_bss *bss)
 }
 
 
+static int nl80211_wowlan_enable(struct i802_bss *bss)
+{
+	struct nl_msg *msg;
+	struct nlattr *wowlan;
+	struct wpa_driver_nl80211_data *drv = bss->drv;
+	int ret = -1;
+
+	msg = nlmsg_alloc();
+	if (!msg)
+		return -ENOMEM;
+
+	wpa_printf(MSG_DEBUG, "nl80211: Enable wowlan");
+	nl80211_cmd(drv, msg, 0, NL80211_CMD_SET_WOWLAN);
+
+	NLA_PUT_U32(msg, NL80211_ATTR_IFINDEX, bss->ifindex);
+	wowlan = nla_nest_start(msg, NL80211_ATTR_WOWLAN_TRIGGERS);
+	if (!wowlan) {
+		wpa_printf(MSG_DEBUG, "nl80211: nla_nest_start error");
+		goto nla_put_failure;
+	}
+	nla_nest_end(msg, wowlan);
+
+	ret = send_and_recv_msgs(drv, msg, NULL, NULL);
+
+	if (ret) {
+			wpa_printf(MSG_DEBUG, "nl80211: enable wowlan on %d"
+				   "err=%d (%s)", bss->ifindex, ret, strerror(-ret));
+	}
+
+	return ret;
+
+nla_put_failure:
+	nlmsg_free(msg);
+	return ret;
+}
+
+
 /**
  * wpa_driver_nl80211_init - Initialize nl80211 driver interface
  * @ctx: context to be used when calling wpa_supplicant functions,
@@ -3694,6 +3731,8 @@ static void * wpa_driver_nl80211_init(void *ctx, const char *ifname,
 		drv->in_interface_list = 1;
 	}
 
+	nl80211_wowlan_enable(bss);
+
 	return bss;
 
 failed:
-- 
1.7.4.1

